src/xrpld/rpc/handlers/DepositAuthorized.cpp

```diff
@@ -88,23 +90,102 @@ doDepositAuthorized(RPC::JsonContext& context)
-    bool depositAuthorized{true};
-    if (srcAcct != dstAcct)
+    bool const reqAuth =
+        (sleDest->getFlags() & lsfDepositAuth) && (srcAcct != dstAcct);
+    bool const credentialsPresent = params.isMember(jss::credentials);
+
+    std::set<std::pair<AccountID, Slice>> sorted;
+    std::vector<std::shared_ptr<SLE const>> lifeExtender;
+    if (credentialsPresent)
     {
-        // Check destination for the DepositAuth flag.  If that flag is
-        // not set then a deposit should be just fine.
-        if (sleDest->getFlags() & lsfDepositAuth)
+        auto const& creds(params[jss::credentials]);
+        if (!creds.isArray() || !creds)
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::credentials,
+                    "is non-empty array of CredentialID(hash256)"));
+        }
+        else if (creds.size() > maxCredentialsArraySize)
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::credentials, "array too long"));
+        }
+
+        lifeExtender.reserve(creds.size());
+        for (auto const& jo : creds)
         {
-            // See if a preauthorization entry is in the ledger.
-            auto const sleDepositAuth =
-                ledger->read(keylet::depositPreauth(dstAcct, srcAcct));
-            depositAuthorized = static_cast<bool>(sleDepositAuth);
+            if (!jo.isString())
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            uint256 credH;
+            auto const credS = jo.asString();
+            if (!credH.parseHex(credS))
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            std::shared_ptr<SLE const> sleCred =
+                ledger->read(keylet::credential(credH));
+            if (!sleCred)
+            {
+                RPC::inject_error(
+                    rpcBAD_CREDENTIALS, "credentials don't exist", result);
+                return result;
+            }
+
+            if (!(sleCred->getFlags() & lsfAccepted))
+            {
+                RPC::inject_error(
+                    rpcBAD_CREDENTIALS, "credentials aren't accepted", result);
+                return result;
+            }
+
+            if (credentials::checkExpired(
+                    sleCred, ledger->info().parentCloseTime))
+            {
+                RPC::inject_error(
+                    rpcBAD_CREDENTIALS, "credentials are expired", result);
+                return result;
+            }
+
```

Comment:
don't we need to verify that the `(*sleCred)[sfSubject]` matches the `srcAcct` ?

Reply:
it is verified 
[here](https://github.com/XRPLF/rippled/pull/5103/files#diff-625db95c5154eaa0cd388d35a7e02117b65bc71b0b6479d0c4157f265ffacafdR233)
```
    if (!requireAuth || (src == dst))
        return tesSUCCESS;
```

Reply:
added check and test